src/test/java/io/vertx/core/ContextTest.java

```diff
@@ -514,29 +516,37 @@ private void checkDuplicate(ContextInternal ctx, ContextInternal duplicated) thr
 
   @Test
   public void testDuplicateWorkerConcurrency() throws Exception {
-    ContextInternal ctx = createWorkerContext();
-    ContextInternal dup1 = ctx.duplicate();
-    ContextInternal dup2 = ctx.duplicate();
-    CyclicBarrier barrier = new CyclicBarrier(3);
-    dup1.runOnContext(v -> {
-      assertTrue(Context.isOnWorkerThread());
-      try {
-        barrier.await(10, TimeUnit.SECONDS);
-      } catch (Exception e) {
-        fail(e);
-      }
-    });
-    dup2.runOnContext(v -> {
-      assertTrue(Context.isOnWorkerThread());
-      try {
-        barrier.await(10, TimeUnit.SECONDS);
-      } catch (Exception e) {
-        fail(e);
-      }
-    });
-    barrier.await(10, TimeUnit.SECONDS);
+    testDuplicateWorkerConcurrency((ctx, task) -> ctx.runOnContext(v -> task.run()));
+    testDuplicateWorkerConcurrency((ctx, task) -> ctx.execute(v -> task.run()));
+    testDuplicateWorkerConcurrency((ctx, task) -> ctx.execute(null, v -> task.run()));
+    testDuplicateWorkerConcurrency(ContextInternal::execute);
+    testDuplicateWorkerConcurrency((ctx, task) -> ctx.emit(v -> task.run()));
+    testDuplicateWorkerConcurrency((ctx, task) -> ctx.emit(null, v -> task.run()));
+  }
+
+  private void testDuplicateWorkerConcurrency(BiConsumer<ContextInternal, Runnable> task) throws Exception {
+    ContextInternal worker = ((VertxInternal)vertx).createWorkerContext();
+    ContextInternal[] contexts = new ContextInternal[] { worker.duplicate(), worker.duplicate()};
+    waitFor(contexts.length);
+    AtomicBoolean owner = new AtomicBoolean();
+    CountDownLatch latch = new CountDownLatch(contexts.length);
+    for (ContextInternal context : contexts) {
+      task.accept(context, () -> {
+        try {
+          assertTrue(owner.compareAndSet(false, true));
+          Thread.sleep(200);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+        } finally {
+          owner.set(false);
+        }
+        latch.countDown();
```

Comment:
This is a minor but in if something goes wrong the latch is never decremented, I usually decrement required `latch`es in the finally block just in case, that way the main coordinating thread will be able to progress in any case.

Reply:
Actually, now I remember why I have 2 try blocks, here is how I would had that code:
```
try {
  assertTrue(owner.compareAndSet(false, true));
  try {
    Thread.sleep(200);
  } catch (InterruptedException e) {
    Thread.currentThread().interrupt();
  } finally {
   // if we reach this finally the owner was successfully set to true by this thread
    owner.set(false);
  }
} finally {
  // Regardless of success or any exception always decrement the latch
  latch.countDown();
}
```